spec/serialization_spec.rb

```diff
@@ -43,6 +45,505 @@ def self.name
+              render
+            }.to raise_error(Dry::Types::ConstraintError)
+          end
+        end
+      end
+
+      context 'when hash' do
+        before do
+          resource.attribute :age, :hash
+        end
+
+        it 'works' do
+          PORO::Employee.create(age: { foo: 'bar' })
+          render
+          expect(attributes['age']).to eq({ 'foo' => 'bar' })
+        end
+
+        context 'when cannot coerce' do
+          before do
+            PORO::Employee.create(age: [:foo, :bar])
+          end
+
+          it 'raises error' do
+            expect {
+              render
+            }.to raise_error(Dry::Types::ConstraintError)
+          end
+        end
+      end
+
+      context 'when array' do
+        before do
+          resource.attribute :age, :array
+        end
+
+        it 'works' do
+          PORO::Employee.create(age: [1, 2, 3])
+          render
+          expect(attributes['age']).to eq([1, 2, 3])
+        end
+
+        it 'applies basic to_json conversion of elements' do
+          time = Time.parse('01-06-2018 4:36pm')
+          PORO::Employee.create(age: [time])
+          render
+          expect(attributes['age']).to eq(["2018-06-01 16:36:00 -0400"])
+        end
+
+        # If we did Array(value), you'd get something incorrect
+        # for hashes
+        it 'raises error on single values' do
+          PORO::Employee.create(age: 1)
+          expect {
+            render
+          }.to raise_error(Dry::Types::ConstraintError)
+        end
+
+        context 'when cannot coerce' do
+          before do
+            PORO::Employee.create(age: { foo: 'bar' })
+          end
+
+          it 'raises error' do
+            expect {
+              render
+            }.to raise_error(Dry::Types::ConstraintError)
+          end
+        end
+      end
+
+      # test for all array_of_*
+      context 'when array_of_integers' do
+        before do
+          resource.attribute :age, :array_of_integers
+        end
+
+        it 'works' do
+          PORO::Employee.create(age: [1, 2, 3])
+          render
+          expect(attributes['age']).to eq([1, 2, 3])
+        end
+
+        it 'applies basic coercion of elements' do
+          PORO::Employee.create(age: ['1', '2', '3'])
+          render
+          expect(attributes['age']).to eq([1, 2, 3])
+        end
+
+        # If we did Array(value), you'd get something incorrect
+        # for hashes
+        it 'raises error on single values' do
+          PORO::Employee.create(age: 1)
+          expect {
+            render
+          }.to raise_error(Dry::Types::ConstraintError)
+        end
+
+        context 'when cannot coerce' do
+          before do
+            PORO::Employee.create(age: { foo: 'bar' })
+          end
+
+          it 'raises error' do
+            expect {
+              render
+            }.to raise_error(Dry::Types::ConstraintError)
+          end
+        end
+      end
+
+      context 'when custom type' do
```

Comment:
Would test `array_of_custom` as well.

Reply:
That would require that if you did `Types[:custom] = MyType` we automatically create `array_of_customs` underneath the hood. I'm not sure I'm in favor of the implicit behavior when you're already customizing.

Reply:
Sorry, I misunderstood what was happening here.  I thought `array_of_custom` was supported but not tested. All good then.